xstream/src/test/com/thoughtworks/acceptance/AbstractAcceptanceTest.java

```diff
@@ -143,8 +143,7 @@ protected void assertObjectsEqual(final Object expected, final Object actual) {
             } else {
                 // assertEquals(expected.getClass(), actual.getClass());
                 if (!expected.equals(actual)) {
-                    assertEquals("Object deserialization failed", "DESERIALIZED OBJECT\n" + xstream.toXML(expected),
-                        "DESERIALIZED OBJECT\n" + xstream.toXML(actual));
+                    assertEquals(xstream.toXML(expected) + " vs. " + xstream.toXML(actual), expected, actual);
```

Comment:
CI build shows 90+ failing tests, and they all appear to be caused by this change. Perhaps some of the other converters are lossy as well, and some of the failure appear to be for objects which do not have `hashCode` and `equals` methods defined, so the comparison fails because we have distinct references before and after deserialization:

```
junit.framework.AssertionFailedError: 
<camera name="Rebel 350">
  <n>
    <name>foo</name>
  </n>
</camera> vs. <camera name="Rebel 350">
  <n>
    <name>foo</name>
  </n>
</camera> expected:<com.thoughtworks.acceptance.AttributeTest$Camera@50d640dd> but was:<com.thoughtworks.acceptance.AttributeTest$Camera@447bf09d>
```

Also one small nit, I would change the spacing to make the failure messages a bit easier to read:
```suggestion
                    assertEquals(xstream.toXML(expected) + "\n  vs.\n" + xstream.toXML(actual), expected, actual);
```

Reply:
Right, I guess this needs to be backed out, but comments inserted to explain that the objects are considered “equal” not in the usual Java sense.

Reply:
By the way I tried to salvage the benefit via

```java
if (!expected.equals(actual)) {
    try {
        if (expected.getClass().getMethod("equals", Object.class).getDeclaringClass() == Object.class) {
            assertEquals("Object deserialization failed", "DESERIALIZED OBJECT\n" + xstream.toXML(expected),
                "DESERIALIZED OBJECT\n" + xstream.toXML(actual));
        } else {
            assertEquals(xstream.toXML(expected) + "\nvs.\n" + xstream.toXML(actual), expected, actual);
        }
    } catch (NoSuchMethodException x) {
        fail(x.toString()); // should be impossible
    }
}
```

but there are still a few failures. I think that is because of _another_ previously unreported lossiness in `StackTraceElementConverter`, which can be seen by this failing test:

```java
public void testIncludesNativeMethods2() {
    final StackTraceElement trace = factory.element("com.blah.SomeClass", "someMethod", "SomeClass.java", -2);
    final String expectedXml = "<trace>com.blah.SomeClass.someMethod(Native Method)</trace>";
    assertBothWays(trace, expectedXml);
}
```

This is not merely theoretical: the JVM actually does construct `StackTraceElement`s for native methods this way. In other words, even though the `toString` form just says `(Native Method)`, internally it is actually tracking a specific `fileName`. Thus, round-tripping a `Throwable` through XStream results in a loss of information whenever native methods are used (quite common). This seems to fix some problems but not others, and seems fragile anyway:

```diff
         public StackTraceElement nativeMethodElement(String declaringClass, String methodName) {
-            return create(declaringClass, methodName, "Native Method", -2);
+            return create(declaringClass, methodName, declaringClass.replaceFirst("^.+[.]", "") + ".java", -2);
         }
```

Reply:
Well, the question is, if this was really lossy for native methods, when the code was written 15 years ago ;-)